### ISSUE TITLE
Infinite recursion when serialized object has a `toJSON` method

### ISSUE THREAD

**Issue by canibanoglu** [WebReflection/circular-json#23]
Hello,

I have come across a problem while using this package in an ExpressJS project. I'm sending the response with `res.json(circularObject)` which internally calls `JSON.stringify`. Since I don't want to explicitly call the `stringify` function provided by this library, I have included a `toJSON` function into my object. 

As per [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#toJSON()_behavior) `JSON.stringify` will call the `toJSON` function if the serialized object has a property named `toJSON` and that property is a function. 

`circular-json`'s stringify function also calls `JSON.stringify` and this results in an infinite loop. A simple check for this condition and accordingly adjusting the `stringifyRecursion` method should resolve the issue. If you are of the same opinion, I can have a go at it and send you a PR.

Kind regards,

Can

**Comment by WebReflection** [comment]
Hi @canibanoglu 

I'm not sure I understand your concern ... so, `CircularJSON` is exactly a `JSON.stringify` wrap that uses a transformer to grant circular representation.

If you have object returning different version of themselves I don't think you should use this library because you know already there won't be circular data.

If I am not getting it, a little example to understand what's exactly the issue would be great.

Thanks

**Comment by canibanoglu** [comment]
Ah, but I know that there will be circular data. 

I won't paste the actual code here for brevity but here's what I'm trying to do.

```js
var cycle = require("circular-json");

function Graph() {
  this.edges = {};
  this.vertices = {};
}

Graph.prototype.toJSON = function() {
  return cycle.stringify(this);
}

function Edge() {
  this.id = "42";
  this.start = null;  // This will be a Vertex instance
  this.end = null;  // This will be a Vertex instance
  this.properties = {};
}

function Vertex() {
  this.id = "42";
  this.outgoingEdges = [];  // There will be many edges here
  this.incomingEdges = [];  // There will be many edges here
  this.properties = {};
}
```

As you can see, when I add an edge and two corresponding vertices to my graph object, I will have circular data. I need to transport this data from the server to the client in JSON format. This is exactly why I need to use `circular-json`.

Here's how I send that data to client in ExpressJS:

```js
router.get('/sidiousEndpoint', function(req, res, next) {
  // Let's say that I have received a Graph object full of data from somewhere
  res.json(graphObject);
});
```

ExpressJS's `res.json` practically does the following: 
 * Serialize the object by calling `JSON.stringify`
 * Send the response by calling `res.send`

I know that I could always do the following without running into the problem I have described in my original comment:

```js
router.get('/sidiousEndpoint', function(req, res, next) {
  // Again, I got a Graph object from somewhere
  res.send(cycle.stringify(graphObject));
});
```

But I believe that this approach will lead to a lot of (in my opinion, unnecessary) calls to `cycle.stringify`, at least one for every endpoint handler from which I return a `Graph` object.

To that end, I opted make use of `JSON.stringify`'s default behaviour of calling `toJSON` if it is present in the serialized object. 

I hope that this clarifies the problem. 

Kind regards,

Can

**Comment by WebReflection** [comment]
Hey @canibanoglu I think you misunderstood what's the `toJSON` purpose.

Example:
```js
JSON.stringify({toJSON:function(){return '{"no":"pe"}'}});
""{\"no\":\"pe\"}""
```

You see that output? That is because you don't return a string with a `toJSON` method, you need to return an object or whatever value you want to represent as JSON.

Example
```js
JSON.stringify({toJSON:function(){return [1,2,3]}});
"[1,2,3]"
```

Your infinite loop has nothing to do with this project, it would have been exactly the same using `function toJSON() { return JSON.stringify(this) }` ... do you understand where's the issue?


**TL;DR** unless you really need to add anything interesting to those instances, don't use `toJSON`

If you need to add something during serialization, clone own properties only and return that object instead.

**Comment by canibanoglu** [comment]
Thank you for the quick reply. I fully understand your point and yeah, with my method the serialized form is wrong. Bah, I'll need to find another way of doing this. 

Thank you very much again for the help and pointers.

**Comment by WebReflection** [comment]
in case you need to clone/merge/copy objects maybe this post can help you:
https://www.webreflection.co.uk/blog/2015/10/06/how-to-copy-objects-in-javascript

There is also a cloner utility explained and usable through npm:
https://github.com/WebReflection/cloner
